aten/src/ATen/native/mps/operations/BinaryKernel.mm

```diff
@@ -0,0 +1,199 @@
+  constant T* other = (constant T*)((constant uint8_t*)other_ + offsets[tid].z);
+
+  *out = fmin(*input, *other);
+}
+
+#define REGISTER_FMAX_OP(DTYPE)                       \
+template                                               \
+[[host_name("fmax_" #DTYPE)]]                         \
+kernel void fmax<DTYPE>(                  \
+  constant void     * input_        [[buffer(0)]],     \
+  constant void     * other_        [[buffer(1)]],     \
+  device   void     * out_          [[buffer(2)]],     \
+  constant uint3    * offsets       [[buffer(3)]],     \
+  uint tid [[thread_position_in_grid]]);
+
+#define REGISTER_FMIN_OP(DTYPE)                       \
+template                                               \
+[[host_name("fmin_" #DTYPE)]]                         \
+kernel void fmin<DTYPE>(                  \
+  constant void     * input_        [[buffer(0)]],     \
+  constant void     * other_        [[buffer(1)]],     \
+  device   void     * out_          [[buffer(2)]],     \
+  constant uint3    * offsets       [[buffer(3)]],     \
+  uint tid [[thread_position_in_grid]]);
+
+REGISTER_FMAX_OP(float);
+REGISTER_FMAX_OP(half);
+REGISTER_FMIN_OP(float);
+REGISTER_FMIN_OP(half);
+
+)BINARY_METAL";
+
+using namespace mps;
+
+static id<MTLLibrary> compileBinaryOpsLibrary(id<MTLDevice> device) {
+  static id<MTLLibrary> binaryLibrary = nil;
+  if (binaryLibrary) {
+    return binaryLibrary;
+  }
+
+  NSError *error = nil;
+  MTLCompileOptions *options = [[MTLCompileOptions new] autorelease];
+  [options setLanguageVersion: MTLLanguageVersion2_3];
+  binaryLibrary  = [device newLibraryWithSource:[NSString stringWithCString: METAL_BINARY encoding:NSASCIIStringEncoding]
+                                       options:options
+                                         error:&error];
+  TORCH_CHECK(binaryLibrary, "Failed to create metal binary library, error: ", [[error description] UTF8String]);
+  return binaryLibrary;
+}
+
+static id<MTLComputePipelineState> binaryPipelineState(id<MTLDevice> device, const std::string& kernel) {
+  static std::unordered_map<std::string, id<MTLComputePipelineState>> psoCache;
+  id<MTLComputePipelineState> pso = psoCache[kernel];
+  if (pso) {
+    return pso;
+  }
+
+  NSError* error = nil;
+  id<MTLLibrary> binaryLib = compileBinaryOpsLibrary(device);
+  id<MTLFunction> binaryFunc = [binaryLib newFunctionWithName:[NSString stringWithUTF8String:kernel.c_str()]];
+  TORCH_CHECK(binaryFunc, "Failed to create function state object for: ", kernel);
+  pso = [device newComputePipelineStateWithFunction:binaryFunc error:&error];
+  TORCH_CHECK(pso, "Failed to created pipeline state object, error: ", [[error description] UTF8String]);
+
+  psoCache[kernel] = pso;
+  return pso;
+}
+
+void fmax_fmin_mps_impl(TensorIteratorBase& iter, const std::string max_min) {
+  TORCH_CHECK(iter.common_dtype() != at::kDouble, "float64 is not supported on MPS");
```

Comment:
Hi,
May i know why assert double datatype here? Is it because the Apple GPU doesn't support double precision?
BTW, why doesn't unary op have such assertion?
Thank you.

Reply:
Yes it doesn't support double precision AFAIK. As for why unary ops don't have such assertion, I think they're just missed. But in general, you're not allowed to create an fp64 MPS tensor nor allowed to type cast / move a fp64 tensor to MPS.

Reply:
Thank you for help. I find the code below,
https://github.com/pytorch/pytorch/blob/71d020262793542974cf13b30f2a9099773f015c/aten/src/ATen/native/mps/OperationUtils.mm#L83-L85
and
https://github.com/pytorch/pytorch/blob/71d020262793542974cf13b30f2a9099773f015c/aten/src/ATen/native/mps/operations/BitwiseOps.mm#L95-L108
Any FP64 data type will throw error in these MPS utilities.

BTW, here code is used to register the supported datatype for the fmax kernel, right? Does it mean only the half and the float32 kernels will be precompiled when AOT compiling.
https://github.com/pytorch/pytorch/blob/71d020262793542974cf13b30f2a9099773f015c/aten/src/ATen/native/mps/operations/BinaryKernel.mm#L122-L123

Thank you.